sklearn/neighbors/classification.py

```diff
@@ -154,21 +156,38 @@ def predict(self, X):
         n_samples = X.shape[0]
         weights = _get_weights(neigh_dist, self.weights)
 
-        y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
-        for k, classes_k in enumerate(classes_):
-            if weights is None:
-                mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
-            else:
-                mode, _ = weighted_mode(_y[neigh_ind, k], weights, axis=1)
+        if self.outputs_2d_ == 'sparse':
+            y_pred_sparse_multilabel = []
 
-            mode = np.asarray(mode.ravel(), dtype=np.intp)
-            y_pred[:, k] = classes_k.take(mode)
+            for k, classes_k in enumerate(classes_):
+                mode = self._mode(_y[neigh_ind, k].toarray(), weights)
+                # `csr_matrix(mode).T` is a CSC matrix with a single column
+                # so that it's fast to hstack it with the other columns
+                y_pred_sparse_multilabel.append(csr_matrix(mode).T)
```

Comment:
Yes, your proposal is faster.
```Python
import scipy.sparse as sp
import numpy as np

f1 = lambda x: sp.csr_matrix(x).T
f2 = lambda x: sp.csc_matrix(x.reshape(-1,1))
data = [np.random.randint(0,2,1000) for _ in range(10000)]

for f in [f1, f2]:
    %timeit sp.hstack([f(d) for d in data])
```
```
1 loop, best of 3: 3.74 s per loop
1 loop, best of 3: 3.1 s per loop
```

But it brakes some test. I'll try to figure it out.